// @glimmer/util/dist/commonjs/es5/lib/debug-to-string.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _env = require('@glimmer/env');

var debugToString = void 0;
if (_env.DEBUG) {
    var getFunctionName = function getFunctionName(fn) {
        var functionName = fn.name;
        if (functionName === undefined) {
            var match = Function.prototype.toString.call(fn).match(/function (\w+)\s*\(/);
            functionName = match && match[1] || '';
        }
        return functionName.replace(/^bound /, '');
    };
    var getObjectName = function getObjectName(obj) {
        var name = void 0;
        var className = void 0;
        if (obj.constructor && obj.constructor !== Object) {
            className = getFunctionName(obj.constructor);
        }
        if ('toString' in obj && obj.toString !== Object.prototype.toString && obj.toString !== Function.prototype.toString) {
            name = obj.toString();
        }
        // If the class has a decent looking name, and the `toString` is one of the
        // default Ember toStrings, replace the constructor portion of the toString
        // with the class name. We check the length of the class name to prevent doing
        // this when the value is minified.
        if (name && name.match(/<.*:ember\d+>/) && className && className[0] !== '_' && className.length > 2 && className !== 'Class') {
            return name.replace(/<.*:/, '<' + className + ':');
        }
        return name || className;
    };
    var getPrimitiveName = function getPrimitiveName(value) {
        return String(value);
    };
    debugToString = function debugToString(value) {
        if (typeof value === 'function') {
            return getFunctionName(value) || '(unknown function)';
        } else if (typeof value === 'object' && value !== null) {
            return getObjectName(value) || '(unknown object)';
        } else {
            return getPrimitiveName(value);
        }
    };
}
exports.default = debugToString;
