// @glimmer/util/dist/modules/es5/index.js
export { EMPTY_ARRAY } from './lib/array-utils';
export { default as assert, deprecate } from './lib/assert';
export { dict, DictSet, isDict, isObject, StackImpl as Stack } from './lib/collections';
export * from './lib/destroy';
export * from './lib/dom';
export { ensureGuid, initializeGuid } from './lib/guid';
export { isSerializationFirstNode, SERIALIZATION_FIRST_NODE_STRING } from './lib/is-serialization-first-node';
export * from './lib/lifetimes';
export { EMPTY_SLICE, LinkedList, ListNode, ListSlice } from './lib/list-utils';
export { assign, fillNulls, values } from './lib/object-utils';
export * from './lib/platform-utils';
export * from './lib/string';
export * from './lib/immediate';
export { default as debugToString } from './lib/debug-to-string';
export function assertNever(value) {
    var desc = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'unexpected unreachable branch';

    console.log('unreachable', value);
    console.trace(desc + ' :: ' + JSON.stringify(value) + ' (' + value + ')');
}
