// @glimmer/syntax/dist/commonjs/es5/lib/traversal/path.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
var _createClass = function () {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);
        }
    }return function (Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;
    };
}();

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

var Path = function () {
    function Path(node) {
        var parent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
        var parentKey = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

        _classCallCheck(this, Path);

        this.node = node;
        this.parent = parent;
        this.parentKey = parentKey;
    }

    Path.prototype.parents = function parents() {
        var _this = this,
            _ref;

        return _ref = {}, _ref[Symbol.iterator] = function () {
            return new PathParentsIterator(_this);
        }, _ref;
    };

    _createClass(Path, [{
        key: "parentNode",
        get: function get() {
            return this.parent ? this.parent.node : null;
        }
    }]);

    return Path;
}();

exports.default = Path;


var PathParentsIterator = function () {
    function PathParentsIterator(path) {
        _classCallCheck(this, PathParentsIterator);

        this.path = path;
    }

    PathParentsIterator.prototype.next = function next() {
        if (this.path.parent) {
            this.path = this.path.parent;
            return { done: false, value: this.path };
        } else {
            return { done: true, value: null };
        }
    };

    return PathParentsIterator;
}();
